python/lsst/ts/tunablelaser/read_serial_temp_scanner.py

```diff
@@ -0,0 +1,336 @@
+import pigpio
+import serial
+from lsst.ts import tcpip
+
+FAN_ON = 1
+FAN_OFF = 0
+
+
+def execute_serial_temperature_scanner():
+    asyncio.run(SerialTemperatureScanner.amain(index=None))
+
+
+class SerialTemperatureScanner(tcpip.OneClientServer):
+    """This is the class that implements the Serial Temperature Scanner script.
+
+    Parameters
+    ----------
+    logger : `logging.Logger`
+        logger object
+    port : `int`, optional
+        port that the server will be hosted on, default 1883
+    host : `str`, optional
+        IP the server will be hosted on, default tcpip.DEFAULT_LOCALHOST
+    encoding : `str`, optional
+        Encoding used for the packets
+    terminator: `bytes`, optional
+        terminating character used for packets
+    sample_wait_time: `int`, optional
+        time to wait between getting temperature samples
+    serial: `serial` or `None`, optional
+        serial object that the temperature scanner device is connected
+    temperature_windows: `int`, optional
+        Amount of temperature windows to average for rolling avg window
+
+    """
+
+    def __init__(
+        self,
+        logger: logging.Logger,
+        port: int | None = 1884,
+        host: str | None = tcpip.DEFAULT_LOCALHOST,
+        encoding: str = tcpip.DEFAULT_ENCODING,
+        terminator: bytes = tcpip.DEFAULT_TERMINATOR,
+        sample_wait_time: int = 5,
+        serial=None,
+        temperature_windows: int = 8,
+    ):
+        super().__init__(
+            log=logger,
+            port=port,
+            host=host,
+            connect_callback=None,
+            monitor_connection_interval=0,
+            name="",
+            encoding=encoding,
+            terminator=terminator,
+        )
+
+        self.serial = serial
+        self.sensor_dict = {}
+        self.fileptr = None
+        if logger is None:
+            logging.basicConfig(
+                format="%(asctime)s %(levelname)-8s %(message)s",
+                level=logging.INFO,
+                datefmt="%Y-%m-%d %H:%M:%S",
+            )
+
+            logger = logging.getLogger(__name__)
+            logger.propagate = True
+            self.logger = logger
+        else:
+            self.logger = logger
+        self.sample_wait_time = sample_wait_time
+
+        # Fan sensor
+        self.fan_sensor = ""
+        self.fan_gpio = None
+        self.fan_turn_on_temp = 0
+        self.fan_turn_off_temp = 0
+
+        self.latest_data = {sensor_name: 0 for sensor_name in self.sensor_dict}
+        self.rolling_temperature = [0 for _ in range(temperature_windows)]
+
+        self.pi = pigpio.pi()
+
+        self.configured = False
+        self.first_run = True
+        self.encoding = encoding
+        self.pending_messages = []
+        self.port = port
+        self.host = host
+
+        # TODO remove this debug stuff
+        self.header = None
+
+        self.config()
+
+    async def amain(self):
+        self.serial_temperature_task()
```

Comment:
I don't think this is going to work. 

When you call this method, it will block the async event loop and all the async tasks in the background will freeze. 

If the calls you make to read and write data are synchronous you will have to wrap them and run them on an executor. 


See https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor for more details.